File: leetcode/no_regu/L8.scala
// 数字计算操作，解耦数据溢出
object L8 {
	def main(args: Array[String]): Unit = {
  assert(myAtoi("-0014") == -14)
  assert(myAtoi("2147483647") == 2147483647)
  assert(myAtoi("   -2147483648") == -2147483648)
  assert(myAtoi("a 123") == 0)
  assert(myAtoi("1.d") == 1)
}
/**
 * 1. 无效空格直接用trim
 * 2. 中间状态用正数存储时，遇到溢出直接返回最值
 * 3. 不必关注溢出，可以提前避免
 */
def myAtoi(s1: String): Int = {
  val s = s1.trim
  var isPositive = true
  var i, res = 0
  if (i < s.length && (s(i) == '+' || s(i) == '-')) {
    if (s(i) == '-') isPositive = false
    i += 1
  }
  while (i < s.length && isNumber(s(i))) {
    val add = s(i) - '0'
    if (isPositive && (res > Int.MaxValue/10 || (res == Int.MaxValue/10 && add >= 7))) return Int.MaxValue
    if (!isPositive && (res > Int.MaxValue/10 || (res == Int.MaxValue/10 && add > 8))) return Int.MinValue

    res = res * 10 + add
    i += 1
  }

  if (!isPositive) res *= -1
  	res
  }

  private def isNumber(c: Char): Boolean = c >= '0' && c <= '9'
}

File: leetcode/no_regu/L8.java
class L8 {
    public int myAtoi(String s) {
        s = s.trim();
        if (s.isEmpty()) return 0;

        int res = 0, i = 0;
        boolean flag = true;
        char tmp = s.charAt(0);
        if (tmp == '+' || tmp == '-') {
            i++;
            if (tmp == '-') flag = false;
        }

        for (; i < s.length() && isNumber(s.charAt(i)); i++) {
            int c = s.charAt(i) - '0';
            
            if (res > Integer.MAX_VALUE / 10 || (res == Integer.MAX_VALUE / 10 && c >= 8)) {
                res = flag ? Integer.MAX_VALUE : Integer.MIN_VALUE;
                return res;
            }
            res = res * 10 + c;
        }

        return res * (flag ? 1 : -1);
    }


    private boolean isNumber(char c) {
        return c >= '0' && c <= '9';
    }
}